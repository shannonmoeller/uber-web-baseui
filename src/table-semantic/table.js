/*
Copyright (c) 2018-2019 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
// @flow

import React, {useMemo, type ComponentType, type Node} from 'react';
import {ChevronDown, ChevronUp} from '../icon/index.js';

import {
  StyledTableContainer,
  StyledTable,
  StyledTableHead,
  StyledTableHeadRow,
  StyledTableHeadCell,
  StyledTableHeadCellIcon,
  StyledTableBody,
  StyledTableBodyRow,
  StyledTableBodyCell,
} from './styled-components.js';

export type SortOrderT = 'ASC' | 'DESC';

export type TableColumnT = {
  key: string,
  header: Node,
  Cell: ComponentType<{row: any}>,
  isLiteral?: boolean,
  isNumeric?: boolean,
  isSortable?: boolean,
};

export type TableHeadCellPropsT = {
  column: TableColumnT,
  sortKey?: ?string,
  sortOrder?: ?SortOrderT,
  onSort?: (string, SortOrderT) => void,
};

export function TableHeadCell(props: TableHeadCellPropsT) {
  const {column, sortKey, sortOrder, onSort, ...rest} = props;
  const {key, header, isSortable} = column;
  const isSorted = isSortable && key === sortKey;

  const icon = isSorted && (
    <StyledTableHeadCellIcon
      $as={sortOrder === 'ASC' ? ChevronUp : ChevronDown}
    />
  );

  function handleSort() {
    if (isSortable && onSort) {
      if (key === sortKey) {
        onSort(key, sortOrder === 'ASC' ? 'DESC' : 'ASC');
      } else {
        onSort(key, 'ASC');
      }
    }
  }

  function handleKeyPress(event) {
    if (event.key === 'Enter') {
      handleSort();
    }
  }

  return (
    <StyledTableHeadCell
      $isSortable={isSortable}
      tabIndex={isSortable ? 0 : null}
      onClick={handleSort}
      onKeyPress={handleKeyPress}
      {...rest}
    >
      {header}
      &nbsp;
      {icon}
    </StyledTableHeadCell>
  );
}

export type TableBodyRowPropsT = {
  columns: Array<TableColumnT>,
  row: any,
};

export function TableBodyRow(props: TableBodyRowPropsT) {
  const {columns, row} = props;

  // Memoize row rendering to improve performance of large tables
  return useMemo(
    () => (
      <StyledTableBodyRow>
        {columns.map(column => (
          <StyledTableBodyCell
            key={column.key}
            $isLiteral={column.isLiteral}
            $isNumeric={column.isNumeric}
          >
            <column.Cell row={row} />
          </StyledTableBodyCell>
        ))}
      </StyledTableBodyRow>
    ),
    [columns, row],
  );
}

export type TablePropsT = {
  columns: Array<TableColumnT>,
  rows: Array<any>,
  sortKey?: ?string,
  sortOrder?: ?SortOrderT,
  onSort?: (string, SortOrderT) => void,
};

export function Table(props: TablePropsT) {
  const {columns, rows, sortKey, sortOrder, onSort} = props;

  return (
    <StyledTableContainer>
      <StyledTable>
        <StyledTableHead>
          <StyledTableHeadRow>
            {columns.map(column => (
              <TableHeadCell
                key={column.key}
                column={column}
                sortKey={sortKey}
                sortOrder={sortOrder}
                onSort={onSort}
              />
            ))}
          </StyledTableHeadRow>
        </StyledTableHead>
        <StyledTableBody>
          {rows.map(row => (
            <TableBodyRow key={row.key} columns={columns} row={row} />
          ))}
        </StyledTableBody>
      </StyledTable>
    </StyledTableContainer>
  );
}
