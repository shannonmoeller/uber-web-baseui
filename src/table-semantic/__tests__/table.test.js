/*
Copyright (c) 2018-2019 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
// @flow

import * as React from 'react';
import {shallow} from 'enzyme';

import {Table} from '../table.js';

describe('Table', () => {
  it('renders expected rows and columns', () => {
    const columns = [
      {
        key: 'foo',
        header: 'Foo',
        Cell: ({row}) => row.foo,
        isNumeric: true,
        isSortable: true,
      },
      {
        key: 'bar',
        header: 'Bar',
        Cell: ({row}) => row.bar,
        isLiteral: true,
      },
      {
        key: 'baz',
        header: 'Baz',
        Cell: ({row}) => row.foo + row.bar,
      },
    ];

    const rows = [
      {key: 'a', foo: 'a', bar: 'a'},
      {key: 'b', foo: 'b', bar: 'b'},
      {key: 'c', foo: 'c', bar: 'c'},
    ];

    expect(
      shallow(
        <Table
          columns={columns}
          rows={rows}
          sortKey="foo"
          sortOrder="ASC"
          onSort={() => {}}
        />,
      ),
    ).toMatchSnapshot();
  });
});
