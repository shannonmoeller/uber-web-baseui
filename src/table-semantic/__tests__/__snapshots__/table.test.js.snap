// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders expected rows and columns 1`] = `
<ForwardRef>
  <ForwardRef>
    <ForwardRef>
      <ForwardRef>
        <TableHeadCell
          column={
            Object {
              "Cell": [Function],
              "header": "Foo",
              "isNumeric": true,
              "isSortable": true,
              "key": "foo",
            }
          }
          key="foo"
          onSort={[Function]}
          sortKey="foo"
          sortOrder="ASC"
        />
        <TableHeadCell
          column={
            Object {
              "Cell": [Function],
              "header": "Bar",
              "isLiteral": true,
              "key": "bar",
            }
          }
          key="bar"
          onSort={[Function]}
          sortKey="foo"
          sortOrder="ASC"
        />
        <TableHeadCell
          column={
            Object {
              "Cell": [Function],
              "header": "Baz",
              "key": "baz",
            }
          }
          key="baz"
          onSort={[Function]}
          sortKey="foo"
          sortOrder="ASC"
        />
      </ForwardRef>
    </ForwardRef>
    <ForwardRef>
      <TableBodyRow
        columns={
          Array [
            Object {
              "Cell": [Function],
              "header": "Foo",
              "isNumeric": true,
              "isSortable": true,
              "key": "foo",
            },
            Object {
              "Cell": [Function],
              "header": "Bar",
              "isLiteral": true,
              "key": "bar",
            },
            Object {
              "Cell": [Function],
              "header": "Baz",
              "key": "baz",
            },
          ]
        }
        key="a"
        row={
          Object {
            "bar": "a",
            "foo": "a",
            "key": "a",
          }
        }
      />
      <TableBodyRow
        columns={
          Array [
            Object {
              "Cell": [Function],
              "header": "Foo",
              "isNumeric": true,
              "isSortable": true,
              "key": "foo",
            },
            Object {
              "Cell": [Function],
              "header": "Bar",
              "isLiteral": true,
              "key": "bar",
            },
            Object {
              "Cell": [Function],
              "header": "Baz",
              "key": "baz",
            },
          ]
        }
        key="b"
        row={
          Object {
            "bar": "b",
            "foo": "b",
            "key": "b",
          }
        }
      />
      <TableBodyRow
        columns={
          Array [
            Object {
              "Cell": [Function],
              "header": "Foo",
              "isNumeric": true,
              "isSortable": true,
              "key": "foo",
            },
            Object {
              "Cell": [Function],
              "header": "Bar",
              "isLiteral": true,
              "key": "bar",
            },
            Object {
              "Cell": [Function],
              "header": "Baz",
              "key": "baz",
            },
          ]
        }
        key="c"
        row={
          Object {
            "bar": "c",
            "foo": "c",
            "key": "c",
          }
        }
      />
    </ForwardRef>
  </ForwardRef>
</ForwardRef>
`;
